File: packages/java/src/test/java/io/greyfoundry/identifold/storage/StorageContractTest.java
package io.greyfoundry.identifold.storage;

import java.util.concurrent.CompletableFuture;

public final class StorageContractTest {
    public static void main(String[] args) {
        StorageAdapter adapter = new StorageAdapter() {
            public CompletableFuture<Boolean> reserve(ReferenceReservation request) {
                return CompletableFuture.completedFuture(true);
            }

            public CompletableFuture<ReferenceMapping> resolve(String reference, String namespace) {
                return CompletableFuture.completedFuture(new ReferenceMapping(
                    "01890f8c-7b2a-7cc3-98b0-112233445566", namespace));
            }

            public CompletableFuture<Long> allocate(SequenceAllocationRequest request) {
                return CompletableFuture.completedFuture(1L);
            }
        };
        var reservation = new ReferenceReservation(
            "01890f8c-7b2a-7cc3-98b0-112233445566", "order", "ORD-0123-4567-89-P");
        if (!adapter.reserve(reservation).toCompletableFuture().join()) throw new AssertionError("reserve");
        if (!adapter.resolve(reservation.reference(), reservation.namespace()).toCompletableFuture().join().machineId()
            .equals(reservation.machineId())) throw new AssertionError("resolve");
        if (adapter.allocate(new SequenceAllocationRequest(
            reservation.machineId(), "receipt", "RCT", null, 4)).toCompletableFuture().join() != 1L) {
            throw new AssertionError("allocate");
        }
    }
}

File: packages/kotlin/src/test/kotlin/io/greyfoundry/identifold/storage/StorageContractTest.kt
package io.greyfoundry.identifold.storage

import kotlin.coroutines.Continuation
import kotlin.coroutines.EmptyCoroutineContext
import kotlin.coroutines.startCoroutine

private class FakeStorage : StorageAdapter {
    override suspend fun reserve(request: ReferenceReservation) = true

    override suspend fun resolve(reference: String, namespace: String) =
        ReferenceMapping("01890f8c-7b2a-7cc3-98b0-112233445566", namespace)

    override suspend fun allocate(request: SequenceAllocationRequest) = 1L
}

private suspend fun exerciseStorageContract() {
    val adapter: StorageAdapter = FakeStorage()
    val reservation = ReferenceReservation(
        "01890f8c-7b2a-7cc3-98b0-112233445566",
        "order",
        "ORD-0123-4567-89-P",
    )
    check(adapter.reserve(reservation))
    check(adapter.resolve(reservation.reference, reservation.namespace)?.machineId == reservation.machineId)
    check(adapter.allocate(SequenceAllocationRequest(
        reservation.machineId, "receipt", "RCT", null, 4,
    )) == 1L)
}

fun main() {
    var failure: Throwable? = null
    ::exerciseStorageContract.startCoroutine(object : Continuation<Unit> {
        override val context = EmptyCoroutineContext
        override fun resumeWith(result: Result<Unit>) {
            failure = result.exceptionOrNull()
        }
    })
    failure?.let { throw it }
}
